tests/favourite-handler.test.ts

```diff
@@ -0,0 +1,198 @@
+import setGovUkSigningJourneyIdMiddleware from "../lambdas/middlewares/session/set-gov-uk-signing-journey-id-middleware";
+import uploadToyItemMiddleware from "../lambdas/middlewares/toy/upload-toy-middleware";
+import { SessionService } from "../lambdas/services/session-service";
+import { CommonConfigKey } from "../lambdas/types/config-keys";
+
+jest.mock("@aws-sdk/lib-dynamodb", () => {
+  const mockPut = jest.fn();
+  mockPut.mockImplementation(() => {
+    return {
+      input: {
+        Item: {
+          sessionId: "test-session-id",
+          toy: "marbles",
+          status: "Authenticated",
+        },
+      },
+    };
+  });
+  return {
+    __esModule: true,
+    ...jest.requireActual("@aws-sdk/lib-dynamodb"),
+    PutCommand: mockPut,
+  };
+}); //  this is so we only mock out the PutCommand
+describe("favourite-handler.ts", () => {
+  let favouriteLambda: FavouriteLambda;
+  let lambdaHandler: middy.MiddyfiedHandler;
+  let logger: jest.MockedObjectDeep<typeof Logger>;
+  let metrics: jest.MockedObjectDeep<typeof Metrics>;
+  let configService: jest.MockedObjectDeep<typeof ConfigService>;
+  let sessionService: jest.MockedObjectDeep<typeof SessionService>;
+  let dynamoDbClient: jest.MockedObjectDeep<typeof DynamoDBDocument>;
+  let putCommand: jest.MockedObjectDeep<typeof PutCommand>;
+
+  const WEEK_IN_SECONDS = 7 * 24 * 60 * 60;
+
+  const mockEvent = {
+    body: JSON.stringify({
+      client_id: "toy-cri",
+      toy: "marbles",
+    }),
+    headers: {
+      ["x-forwarded-for"]: "test-client-ip-address",
+      ["session-id"]: "6b0f3490-db8b-4803-967d-39d77a2ece21",
+    } as APIGatewayProxyEventHeaders,
+  } as APIGatewayProxyEvent;
+
+  const getMockFetch = (code?: number) => {
+    return jest.fn(() =>
+      Promise.resolve({
+        json: () =>
+          Promise.resolve({
+            toy: "marbles",
+          }),
+        status: code ? code : 200,
+      } as Response)
+    );
+  };
+
+  beforeEach(() => {
+    logger = jest.mocked(Logger);
+    metrics = jest.mocked(Metrics);
+    configService = jest.mocked(ConfigService);
+    sessionService = jest.mocked(SessionService);
+    dynamoDbClient = jest.mocked(DynamoDBDocument);
+    putCommand = jest.mocked(PutCommand);
+
+    favouriteLambda = new FavouriteLambda();
+
+    lambdaHandler = middy(favouriteLambda.handler.bind(favouriteLambda))
+      .use(
+        errorMiddleware(logger.prototype, metrics.prototype, {
+          metric_name: "favourite_selected",
+          message: "Favourite lambda error occurred",
+        })
+      )
+      .use(injectLambdaContext(logger.prototype, { clearState: true }))
+      .use(
+        initialiseConfigMiddleware({
+          configService: configService.prototype,
+          config_keys: [CommonConfigKey.SESSION_TABLE_NAME],
+        })
+      )
+      .use(
+        getSessionByIdMiddleware({ sessionService: sessionService.prototype })
```

Comment:
Normally I'd expect when using combinations of middleware I'd expect to maybe have 1 test that uses the actual implementations of the middleware, and then most of the other tests mocking responses.

For example, in the lines above, mocking `initialiseConfigMiddleware` would allow us to return an error, and then ensure that the expected error flow is handled.

Also, as they're separate functions, they should be able to have their own exstensive tests, separate to their usage in the lambda function.

I'm aware that a) these middleware are copied from `common` and b) there's no current pattern for these approaches, so I'm not expecting them to be fixed in this PR.

Reply:
That's totally fair - I'll move some of the tests to a middleware file